### PR TITLE
Specify the kind of the item for E0121

### LINKED ISSUE #86005: The meaning of an "item" in E0121 should be elaborated on
I tried this code:

```rust
fn main() {
    const BAD: [_; 3] = [1u8, 2u8, 3u8];
}
```

I expected to see this happen: an error which explains why this doesn't work with `const` even though it does work with `let`, like what would happen with the code:

```rust
fn main() {
    const BAD = 1u8;
}
```

Instead, this happened: E0121, which doesn't explain that the problem is that const's need to be explicitly typed, but instead refers to an "item signature", which is an unexplained term of art. The explanation in `rustc --explain E0121` does not elaborate on what an "item signature" means either.

### Meta

`rustc --version --verbose`:
Occurs on 1.52.1 stable and on 1.54.0 nightly.

### PR DESCRIPTION
**Pull request by fee1-dead**
Fixes #86005

### CODE REVIEW

File: src/test/ui/did_you_mean/bad-assoc-ty.stderr [hunk 1]
```diff
@@ -135,7 +135,7 @@ help: use type parameters instead
 LL | fn foo<X: K<T, T>, T>(x: X) {}
    |             ^  ^ ^^^
 
-error[E0121]: the type placeholder `_` is not allowed within types on item signatures
+error[E0121]: the type placeholder `_` is not allowed within functions
```
Comment (petrochenkov) [inline src/test/ui/did_you_mean/bad-assoc-ty.stderr:138]: Well, `_` types are certainly allowed within functions.

The old wording was more precise by talking about types in signatures.
I suggest to extend it with the item kind instead:
```rust
"the type placeholder `_` is not allowed within types on {kind} item signatures"
```

File: compiler/rustc_typeck/src/collect.rs [hunk 2]
```diff
@@ -227,11 +228,30 @@ fn reject_placeholder_type_signatures_in_item(tcx: TyCtxt<'tcx>, item: &'tcx hir
...
+        hir::ItemKind::Union(..) => "union",
+        hir::ItemKind::Enum(..) => "enum",
+        hir::ItemKind::TraitAlias(..) => "trait alias",
+        hir::ItemKind::Trait(..) => "trait",
+        hir::ItemKind::Impl(..) => "implementation",
+        hir::ItemKind::Struct(..) => "struct",
+        hir::ItemKind::OpaqueTy(..) => "opaque type",
+        hir::ItemKind::TyAlias(..) => "type alias",
+        _ => return,
+    };
```
Comment (petrochenkov) [inline compiler/rustc_typeck/src/collect.rs:241]: Could you move this to `impl hir::ItemKind` and make the wording compatible with `ast::ItemKind::descr`?

File: src/test/ui/did_you_mean/bad-assoc-ty.stderr (hunk 1 above)
Comment (fee1-dead) [reply src/test/ui/did_you_mean/bad-assoc-ty.stderr:138]: I don't know how "not allowed within types on function(s) item signatures" is going to work, so I guess I should change it to "not allowed within types on item signatures for {kind}/functions"

**Comment by fee1-dead** [comment]
`mingw-check` has timed out. Force pushing amend commit with no changes to retrigger CI.

**Comment by fee1-dead** [comment]
I splitted the single commit into two, because a lot of the changes were in ui tests and not in `compiler/rustc_*`.

The diff of the first commit can be easier to review.

**Comment by petrochenkov** [comment]
Seems fine.
@bors r+

**Review by petrochenkov — APPROVED** [review]

**Comment by fee1-dead** [comment]
@bors r=petrochenkov

I only added a commit to fix builds after #83739.

(I wanted to see if I have the rights :p)

**Pull request merged**